### PR TITLE
luci-app-olsr-viz: migrate to JavaScript-based implementation

### PR DESCRIPTION
**Pull request by Ayushmanwebdeveloper**
image

Hi! I've completed the migration of luci-app-olsr-viz to JavaScript. All the features are functioning just like they did in the old Lua app.

And in the old app, there was an issue of HTTP 413 Content Too Large after running it twice, which resulted in a blank page because the app was saving 130 Cookies at a time as it was saving all the nodes in cookies. I tested it on many routers and was able to reproduce the issue every time.

Screenshot 2023-07-25 204605

So, I removed the saving of the nodes in the cookies, as it was making the app unusable.

Now everything is working normally, and I've fixed the error.

### CODE REVIEW

File: applications/luci-app-olsr-viz/root/usr/share/rpcd/acl.d/luci-app-olsr-viz-unauthenticated.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+	"unauthenticated": {
+		"description": "Grant read access for luci-app-olsr-viz",
+		"read": {
+			"ubus": {
+				"luci-rpc": [
```
Comment (jow-) [inline applications/luci-app-olsr-viz/root/usr/share/rpcd/acl.d/luci-app-olsr-viz-unauthenticated.json:6]: This access permission seems unneeded

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+		}, 15);
+	},
+	render: function () {
+		var viz_res;
+		var self = this;
+		return this.action_olsr_viz()
+			.then(function (result) {
+				viz_res = result.viz_data;
+
+				var cssScript = E('script', {
```
Comment (jow-) [inline applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js:62]: Oh dear, that's some artifact :) You can replace the CSS injection with just:

```javascript
document.querySelector('head').appendChild(E('style', { 'type': 'text/css' }, [
  '.label {color:black;background-color:white}'
]));
```

To load the `olsr-viz.js` script I suggest to inject the `<script>` tag from within the `load` function, then wrap the onload event callback into a promise to defer `render()` until the script is loaded. This way you can subsequently skip the `var script = '<script language="JavaScript1.2" type="text/javascript">viz_setup("RSIFrame", "main","nodes", "edges");viz_update();</script>';` part and simply call `viz_setup(...)` and `viz_update()` synchronously.

Within the `load` function:

```javascript
return new Promise(function(resolveFn, rejectFn) {
    var script = E('script', { 'type': 'text/javascript' });
    script.onload = resolveFn;
    script.onerror = rejectFn;
    script.src = L.resource('olsr-viz.js');
    document.querySelector('head').appendChild(script);
});
```

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+						zoomInput,
+						E('b', { 'title': 'Beschränkt die Anzeige auf eine maximale Hop-Entfernung.' }, '&nbsp;&nbsp;Metrik'),
+						E('a', { 'href': 'javascript:set_maxmetric(maxmetric+1)' }, '+ '),
+						E('a', { 'href': 'javascript:if(0<maxmetric)set_maxmetric(maxmetric-1)' }, '\u2212'),
+						metricInput,
+						E('b', { 'title': 'Schaltet die automatischen Layout-Optimierung ein.' }, '&nbsp;&nbsp;Optimierung'),
+						autoOptimizationCheckbox,
+						E('b', { 'title': 'Zeige Hostnamen an.' }, ' |  Hostnamen'),
+						hostnamesCheckbox,
+						E('a', { 'href': 'javascript:viz_save()', 'title': 'Speichert die aktuellen Einstellungen in einem Cookie.', 'style': 'font-weight:700;' }, '&nbsp;|&nbsp;&nbsp;Speichern'),
```
Comment (jow-) [inline applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js:135]: I suggest to replace all `href: 'javascript:...'` with `href: '#', click: ...`, e.g. `E('a', { 'href': '#', 'click': viz_save, 'title': ...`
(This requires loading `olsr-viz.js` from `load()`

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js (hunk 2 above)
Comment (Ayushmanwebdeveloper) [reply applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js:62]: Okay I'll fix them all, Thanks for the review!

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js (hunk 3 above)
Comment (Ayushmanwebdeveloper) [reply applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js]: I fixed all issues and tested it on my router as well. You can take a look. Thanks
https://github.com/openwrt/luci/compare/3b0a64c10c525d3ff4da2b89c8c05462faf9d549..f0ca73364f747064142ea31a46ea3173b4edef6b

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+	load: function () {
+		var self = this;
+		poll.add(function () {
+			self.render();
+		}, 15);
+		return new Promise(function (resolve, reject) {
+			var script = E('script', { 'type': 'text/javascript' });
+			script.onload = function () {
+				resolve();
```
Comment (jow-) [inline applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js:57]: You can shorten this to just `script.onload = resolve;` - we don't care about the actual return values or arguments, just that the resolve function is invoked.

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+			};
+			script.onerror = function (err) {
+				reject(err);
```
Comment (jow-) [inline applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js:60]: You can shorten this to just `script.onerror = reject;` - we don't care about the actual return values or arguments, just that the resolve function is invoked.

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+		});
+	},
+	render: function () {
+		var viz_res;
+		var self = this;
+		return this.action_olsr_viz()
+			.then(function (result) {
+				viz_res = result.viz_data;
+
+				document.querySelector('head').appendChild(E('style', { 'type': 'text/css' }, [
```
Comment (jow-) [inline applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js:73]: For the sake of consistency it probably makes sense to move this style tag inject to the `load()` function as well.

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+					'.label {color:black;background-color:white}'
+				]));
+
+				var mainDiv = E('div', {
+					'id': 'main',
+					'style': 'width: 100%; height: 93%; border: 1px solid #ccc; margin-left:auto; margin-right:auto; text-align:center; overflow: scroll'
```
Comment (jow-) [inline applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js:79]: Since we're injecting CSS anyway, it might make sense to add another rule to it:

```javascript
document.querySelector('head').appendChild(E('style', { 'type': 'text/css' }, [
    ...,
    '#main { width: 100%; height: 93%; border: 1px solid #ccc; margin-left:auto; margin-right:auto; text-align:center; overflow: scroll }'
]));
```

Then you can drop the longish `style` property here. Since `#main` is a quite generic ID that might collide with existing theme markup, maybe consider renaming to `olsr_viz_main` or use a CSS class instead.

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js (hunk 3 above)
Comment (Ayushmanwebdeveloper) [reply applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js]: I also fixed the script.

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+					'onchange': 'set_scale()'
+				});
+				var metricInput = E('input', {
+					'id': 'maxmetric',
+					'name': 'maxmetric',
+					'type': 'text',
+					'value': '3',
+					'size': '4',
+					'style': 'min-width: unset !important;',
+					'onchange': 'set_maxmetric(this.value)'
```
Comment (jow-) [inline applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js:101]: I believe this can now be `'change': (ev) => set_maxmetric(ev.target.value)`  (note the missing `on` prefix)

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+				});
+				var autoOptimizationCheckbox = E('input', {
+					'id': 'auto_declump',
+					'name': 'auto_declump',
+					'type': 'checkbox',
+					'onchange': 'set_autodeclump(this.checked)',
```
Comment (jow-) [inline applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js:107]: `'change': (ev) => set_autodeclump(ev.target.checked),`

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+					'checked': 'checked'
+				});
+				var hostnamesCheckbox = E('input', {
+					'id': 'show_hostnames',
+					'name': 'show_hostnames',
+					'type': 'checkbox',
+					'onchange': 'set_showdesc(this.checked)',
```
Comment (jow-) [inline applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js:114]: `'change': (ev) => set_showdesc(ev.target.checked),`

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+					'checked': 'checked'
+				});
+
+				var form = E('form', { 'action': '' }, [
+					E('p', {}, [
+						E('b', { 'title': 'Bestimmt die Vergrößerungsstufe.' }, 'Zoom '),
+						E('a', { 'href': '#', 'onclick': 'set_scale(scale+0.1)' }, '+ '),
```
Comment (jow-) [inline applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js:121]: `'click': () => set_scale(scale + 0.1)`

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+						E('a', { 'href': '#', 'onclick': 'set_scale(scale-0.1)' },  '\u2212 '),
```
Comment (jow-) [inline applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js:122]: `'click': () => set_scale(scale - 0.1)`

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+						zoomInput,
+						E('b', { 'title': 'Beschränkt die Anzeige auf eine maximale Hop-Entfernung.' }, '&nbsp;&nbsp;Metrik'),
+						E('a', { 'href': '#', 'onclick': 'set_maxmetric(maxmetric+1)' }, '+ '),
```
Comment (jow-) [inline applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js:125]: `'click': () => set_maxmetric(maxmetric + 1)`

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+						E('a', { 'href': '#', 'onclick': 'if(0<maxmetric)set_maxmetric(maxmetric-1)' }, '\u2212'),
```
Comment (jow-) [inline applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js:126]: `'click': () => set_maxmetric(Math.max(maxmetric, 1) - 1)`

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+						metricInput,
+						E('b', { 'title': 'Schaltet die automatischen Layout-Optimierung ein.' }, '&nbsp;&nbsp;Optimierung'),
+						autoOptimizationCheckbox,
+						E('b', { 'title': 'Zeige Hostnamen an.' }, ' |  Hostnamen'),
+						hostnamesCheckbox,
+						E('a', { 'href': '#', 'onclick': 'viz_save()', 'title': 'Speichert die aktuellen Einstellungen in einem Cookie.', 'style': 'font-weight:700;' }, '&nbsp;|&nbsp;&nbsp;Speichern'),
```
Comment (jow-) [inline applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js:132]: `'click': () => viz_save()` or even just `'click': viz_save` if `viz_save()` ignores its arguments.

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+						E('a', { 'href': '#', 'onclick': 'viz_reset()', 'title': 'Startet das Viz-Skriptprogramm neu.', 'style': 'font-weight:700;' }, '&nbsp;|&nbsp;&nbsp;Zur&uuml;cksetzen')
```
Comment (jow-) [inline applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js:133]: `'click': () => viz_reset()` or even just `'click': viz_reset` if `viz_reset()` ignores its arguments.

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+					])
+				]);
+
+				var debugSpan = E('span', { 'id': 'debug', 'style': 'visibility:hidden;' });
+
+				var vizDiv = E('div', { 'id': 'RSIFrame', 'name': 'RSIFrame', 'style': 'border:0px; width:0px; height:0px; visibility:hidden;' });
+				var vizScript = E('script', { 'type': 'text/javascript' }, 'viz_setup("RSIFrame", "main","nodes", "edges");viz_update();');
```
Comment (jow-) [inline applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js:140]: I suggest to modify `viz_setup()` to take the target div directly instead of a div name. Then simply replace this delayed script instantiation with a synchronous call to `viz_setup(vizDiv, "main", "nodes", "edges"); viz_update();`

and the `vizDiv` instantiation with `var vizDiv = E('div', { 'id': 'RSIFrame', 'name': 'RSIFrame', 'style': 'border:0px; width:0px; height:0px; visibility:hidden;' }, viz_res);

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js (hunk 4 above)
Comment (Ayushmanwebdeveloper) [reply applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js:57]: Yes! Okay Sure

File: applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js (hunk 17 above)
Comment (Ayushmanwebdeveloper) [reply applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js]: I fixed the issues & also modified script file for directly using divs & inputs instead of getting them by id. and I implemented calling`viz_setup(vizDiv, "main", "nodes", "edges"); viz_update();` synchronously.

Comment (Ayushmanwebdeveloper) [reply applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js]: And I will test it for some time more.

Comment (Ayushmanwebdeveloper) [reply applications/luci-app-olsr-viz/htdocs/luci-static/resources/view/olsr-viz/olsr-viz-view.js]: I tested it & fixed all the issues. Thanks for the review.

**Comment by systemcrash** [comment]
my $0.02 - I like this.

**Comment by andibraeu** [comment]
@jow- this migrated app looks good to me

**Comment by jow-** [comment]
Merged, thanks!
